### PR TITLE
Add an option to set lower TTL for query cache within a time limit

### PR DESCRIPTION
**Pull request by codesome**
_technically a draft, but not opened as draft to send out github notifications for comments_

#### What this PR does

This is a temporary solution to https://github.com/grafana/mimir/issues/2275

We allow query-frontend to set a lower TTL for cache entries for the query results falling within a certain time period, i.e. upto X hours back in time. This X is given via a new flag, which will be set to the out-of-order time window in practice.

The lower TTL is hardcoded right now to 10m. Which means upon ingestion of old data (older than max-cache-freshness), those samples will be available for querying again after 10 mins if the old result for that query was already cached.

I am opening this PR to see if I am in the right direction and if this solution is acceptable. I have left a TODO in the code to split the result of the query so that we can selectively set the usual TTL and lower TTL instead of setting lower TTL to the entire result if it overlaps with the period from the new flag.

#### Which issue(s) this PR fixes or relates to

Kind of https://github.com/grafana/mimir/issues/2275, but it doesn't close the issue

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by jesusvazquez — COMMENTED** [review]
I read the code and everything looks good but I'd like to share what I had in mind.

I was thinking that we should make the query frontend cache ttl settings agnostic to our out of order needs or any other related issues that come out of caching for a longer periods of time. 

What do you think about simply allowing to override the default TTL of 7 days to any other value specified by the new configuration option. 

We could also explore the idea of setting a validation check for minimum TTL value i.e 10 minutes.

**Comment by codesome** [comment]
> What do you think about simply allowing to override the default TTL of 7 days to any other value specified by the new configuration option.

That will make it 2 additional configuration in total. I was trying to keep it simple to use so that by the time we start requiring variable TTL, we will (hopefully) have the actual fix for the cache invalidation and we won't require this temporary fix anymore.

**Review by jesusvazquez — APPROVED** [review]
LGTM but i would wait for a review from someone with more context on the query frontend

**Comment by pracucci** [comment]
_I'm reviewing it._

**Review by pracucci — COMMENTED** [review]
👍  No objection from me, but I left a couple of questions.

Other than the few comments / questions, please:
- Mention it experimental in `docs/sources/operators-guide/configure/about-versioning.md`
- Add a CHANGELOG entry

File: pkg/frontend/querymiddleware/results_cache.go
Comment (codesome) [reply pkg/frontend/querymiddleware/results_cache.go:59]: > Have you considered not adding this config at all, and reading the OOO time window setting instead? I feel this setting will make more complicated to use OOO ingestion.

Yes. But since the ooo time window is an ingester config, we were not sure if it a good idea to use it here as well, because what if users do not set ooo time window for query-frontend? But since ooo time window is in limits (without mention of ingesters in the config name anywhere), we can re-use it and add it in docs that the same limit must be supplied for the query-frontend, hence get rid of the additional flag. WDYT?

Comment (pracucci) [reply pkg/frontend/querymiddleware/results_cache.go:59]: We already have several config options which apply to multiple config. It's fine, as far as documented. The best practice of running Mimir is just using `config.yaml` and passing the same config to all components, so that you don't have to worry about which config applies to which component.

Comment (pracucci) [reply pkg/frontend/querymiddleware/results_cache.go:59]: As a reference, look at this config option which must to be set on 3 components:
https://github.com/grafana/mimir/pull/2957/files#diff-b6410588e11858b48f519e0416032fc64c292f87fab23498d0d2b98aed9c62a2R1195-R1200

File: pkg/frontend/querymiddleware/split_and_cache.go
Comment (pracucci) [inline pkg/frontend/querymiddleware/split_and_cache.go:392]: From a very quick look at the code, I can't get if it's always guaranteed extends are sorted by end time. Have you found a proof of it? Otherwise it's probably safer to iterate through all extents.

File: pkg/frontend/querymiddleware/split_and_cache.go
Comment (codesome) [reply pkg/frontend/querymiddleware/split_and_cache.go:390]: The model.Duration here just wraps the time.Duration, see [here](https://github.com/prometheus/common/blob/main/model/time.go#L169-L172). model.Time is the one having milliseconds.

File: pkg/frontend/querymiddleware/split_and_cache.go
Comment (colega) [inline pkg/frontend/querymiddleware/split_and_cache.go:408]: I think that we should make the lower ttl apply only to the extent that is affected by it, otherwise a tenant with OOO enabled will have much higher load after this. Imagine someone rendering 30d graphs all the time, if they query up to now (which is the usual case), right now we cache the previous days for a week, but now we'll be recalculating them every 10 minutes.

I would add a `ValidUntil` timestamp to the `Extent`, and set it to `time.Now().Add(resultsCacheLowerTTL)` for the extents falling into the OOO window. Then when retrieving the cache, we would just filter out the ones that have `extent.ValidUntil > 0 && time.Now().After(extent.ValidUntil)`

Comment (codesome) [reply pkg/frontend/querymiddleware/split_and_cache.go]: Can I use the same key to store different sets of extents? I don't totally know how this cache works, I will need to check. Also, a `ValidUntil` on the extent feels like duplicating what a cache already does.

Comment (colega) [reply pkg/frontend/querymiddleware/split_and_cache.go]: I didn't undertand this question, but I see your later changes and I think you're on the right path.

Comment (colega) [reply pkg/frontend/querymiddleware/split_and_cache.go]: Okay, clarified offline. I thought that all extents were stored under the same key, and that this change would expire multiple days every 30 min. 

Sorry for the misunderstanding. This looks good to me now.

**Review by colega — APPROVED** [review]

**Review by pracucci — APPROVED** [review]
Thanks!

**Pull request merged**
